recommenders/models/ncf/dataset.py

```diff
@@ -1,321 +1,450 @@
+        return {self.col_user: user, self.col_item: item, self.col_rating: rating, self.col_test_batch: test_batch}
+
+    def _init_data(self):
+        # Compile lists of unique users and items, assign IDs to users and items,
+        # and ensure file is sorted by user (and batch index if test set)
+        logger.info("Indexing {} ...".format(self.filename))
+        with self:
+            user_items = []
+            self.item2id, self.user2id = OrderedDict(), OrderedDict()
+            batch_index = 0
+            for _ in self:
+                item = self.row[self.col_item]
+                user = self.row[self.col_user]
+                test_batch = self.row[self.col_test_batch]
+                if not self.end_of_file:
+                    next_user = self.next_row[self.col_user]
+                    next_test_batch = self.next_row[self.col_test_batch]
+                if item not in self.items:
+                    self.item2id[item] = len(self.item2id)
+                user_items.append(item)
+
+                if (next_user != user) or self.next_row is None:
+                    if not self.end_of_file:
+                        if next_user in self.users:
+                            raise ValueError("File {} is not sorted by user".format(self.filename))
+                    self.user2id[user] = len(self.user2id)
+
+                if self.col_test_batch:
+                    if (next_test_batch != test_batch) or self.next_row is None:
+                        if not self.end_of_file:
+                            if next_test_batch < batch_index:
+                                raise ValueError("File {} is not sorted by {}".format(self.filename, self.col_test_batch))
+                        batch_index += 1
+
+            self.batch_indices_range = range(0, batch_index)
+            self.data_len = self.line_num - 1
+
+    def load_data(self, key, by_user=True):
+        """ Load data for a specified user or test batch
+
+            Args:
+                key (int): user or test batch index
+                by_user (bool): load data by usr if True, else by test batch
+
+            Returns:
+                pandas.DataFrame
+        """
+        records = []
+        key_col = self.col_user if by_user else self.col_test_batch
+
+        # fast forward in file to user/test batch
+        while (self.line_num == 0) or (self.row[key_col] != key):
+            if self.end_of_file:
+                raise Exception("User {} not in file {}".format(key, self.filename))
+            next(self)
+
+        # collect user/test batch data
+        while self.row[key_col] == key:
+            row = self.row
+            if self.col_test_batch in row:
+                del row[self.col_test_batch]
+            records.append(row)
+            if not self.end_of_file:
+                next(self)
+            else:
+                break
+
+        return pd.DataFrame.from_records(records)
+
 
 class Dataset(object):
```

Comment:
there is an error in the PR:

```
=================================== FAILURES ===================================
___________________________ test_data_preprocessing ____________________________

python_dataset_ncf = (     userID  itemID  rating  timestamp
299       1      96       1 2018-01-19
655       1      83       3 2018-08-24
...    3 2020-01-28
730      99      55       4 2020-02-20
728      99      11       4 2020-06-25

[399 rows x 4 columns])

    def test_data_preprocessing(python_dataset_ncf):
        train, test = python_dataset_ncf
        data = Dataset(
>           train=train, test=test, n_neg=N_NEG, n_neg_test=N_NEG_TEST, seed=SEED
        )
E       TypeError: __init__() got an unexpected keyword argument 'train'

tests/unit/recommenders/models/test_ncf_dataset.py:21: TypeError
______________________________ test_train_loader _______________________________

python_dataset_ncf = (     userID  itemID  rating  timestamp
299       1      96       1 2018-01-19
655       1      83       3 2018-08-24
...    3 2020-01-28
730      99      55       4 2020-02-20
728      99      11       4 2020-06-25

[399 rows x 4 columns])

    def test_train_loader(python_dataset_ncf):
        train, test = python_dataset_ncf
        data = Dataset(
>           train=train, test=test, n_neg=N_NEG, n_neg_test=N_NEG_TEST, seed=SEED
        )
E       TypeError: __init__() got an unexpected keyword argument 'train'

tests/unit/recommenders/models/test_ncf_dataset.py:45: TypeError
_______________________________ test_test_loader _______________________________

python_dataset_ncf = (     userID  itemID  rating  timestamp
299       1      96       1 2018-01-19
655       1      83       3 2018-08-24
...    3 2020-01-28
730      99      55       4 2020-02-20
728      99      11       4 2020-06-25

[399 rows x 4 columns])

    def test_test_loader(python_dataset_ncf):
        train, test = python_dataset_ncf
>       data = Dataset(train=train, test=test, n_neg=N_NEG, n_neg_test=N_NEG_TEST)
E       TypeError: __init__() got an unexpected keyword argument 'train'
```
see logs: https://github.com/microsoft/recommenders/runs/4944649632?check_suite_focus=true. 

However, I realised that apart from the error, these tests should be in the GPU set, but are in the CPU. So we need to add `@pytest.mark.gpu` to the tests

Reply:
Yes the tests only work for the old code. I am still working on the revised set of tests. I'll be sure to place them in the GPU set.